apollo-router/src/plugins/cache/entity.rs

```diff
@@ -1526,71 +1526,148 @@ fn collect_key_field_sets(
+        match value {
+            // objects have selection sets, so recurse!
+            Value::Object(sub_value) => {
+                if !matches_selection_set(sub_value, &field.selection_set) {
+                    return false;
+                }
+            }
+            // arrays can be used in complex `@key`s, both arrays of scalars (eg, ["a", "b", "c"],
+            // which are handled above when we check whether the selection set is empty and it will be
+            // because scalars have no selection set; we would have matched in virtue of finding
+            // the scalar in the JSON representation) and arrays of objects (eg, [ObjectA, ObjectB,
+            // ObjectC], which we handle below)
+            Value::Array(arr) => {
+                for item in arr {
+                    // arrays of scalars are handled above; so, we know we have an object--if this
+                    // fails, we've hit a bug
+                    let Value::Object(obj) = item else {
+                        tracing::trace!(
+                            "encounterd an array filled with neither scalars or objects while trying to find an entity key match: {item}"
+                        );
+                        return false;
+                    };
+                    if !matches_selection_set(obj, &field.selection_set) {
+                        return false;
+                    }
+                }
+            }
+            // this should never be hit; we should only have JSON objects and arrays
+            other => {
+                tracing::trace!(
+                    "encounterd something other than an array or object while trying to find an entity key match: {other}"
+                );
+                return false;
+            }
         }
     }
     true
 }
 
-// Removes the selection set from `representation` and returns the value corresponding to it.
-// - Returns None if the representation doesn't match the selection set.
+/// Removes the selection set from `representation` and returns the value(s) corresponding
+/// to the selection set. If no match is found, `None` is returned
+///
+/// The primary role of this function is to find the entity we're after so that we can eventually
+/// create a hashed cache key using it
+// WARN: any changes made to this fn should have the same done to the one in
+// response_cache/plugin.rs
 fn take_selection_set(
     representation: &mut serde_json_bytes::Map<ByteString, Value>,
     selection_set: &apollo_compiler::executable::SelectionSet,
 ) -> Option<serde_json_bytes::Map<ByteString, Value>> {
     let mut result = serde_json_bytes::Map::new();
     for field in selection_set.root_fields(&Default::default()) {
-        // Note: field sets can't have aliases.
+        // selection sets are empty when we're at the terminus (leaf) of the query, either
+        // bottoming out in a scalar (happy path) or an object with no fields (sad path--this
+        // _shouldn't_ happen)
```

Comment:
Indeed, this sad path shouldn't happen, since it's been checked in the `match` function above.

However, if we implement my optimization idea where we skip `match` if the entity type has only one `@key` directive. Then, this sad path will happen in this `take` function, too.